### PR TITLE
[eclipse/xtext-xtend#528] Don't request a build when changing compile…

### PR DESCRIPTION
**Pull request by tivervac**
…r settings that don't require a rebuild

Signed-off-by: Titouan Vervack <tivervac@gmail.com>

### CODE REVIEW

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/preferences/AffectsBuilderPredicate.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	public boolean apply(Entry<String, ValueDifference<String>> input) {
+		String key = input.getKey();
+		if (key.startsWith(PREFIX)) {
+			String keyName = key.substring(key.lastIndexOf(PreferenceConstants.SEPARATOR) + 1);
+			switch (keyName) {
+				case OUTPUT_CREATE_DIRECTORY:
+				case OUTPUT_OVERRIDE:
+				case OUTPUT_DERIVED:
+				case OUTPUT_CLEANUP_DERIVED:
+				case OUTPUT_CLEAN_DIRECTORY:
```
Comment (cdietrich) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/preferences/AffectsBuilderPredicate.java]: @szarnekow i am not sure which of these should go where

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/preferences/BuilderConfigurationBlock.java [hunk 2]
```diff
@@ -52,6 +55,8 @@
 public class BuilderConfigurationBlock extends OptionsConfigurationBlock {
 	
 	protected static final String[] BOOLEAN_VALUES = new String[] { IPreferenceStore.TRUE, IPreferenceStore.FALSE };
+
+	private static final Predicate<Entry<String, ValueDifference<String>>> AFFECTS_BUILDER_PREDICATE = new AffectsBuilderPredicate();
```
Comment (cdietrich) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/preferences/BuilderConfigurationBlock.java]: am not sure if we need a factory for that to be better configurable

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/preferences/BuilderConfigurationBlock.java [hunk 3]
```diff
@@ -388,4 +393,8 @@ public String getPropertyPrefix() {
 		return PROPERTY_PREFIX;
 	}
 
+	@Override
+	protected Predicate<Entry<String, ValueDifference<String>>> getAffectsBuilderPredicate() {
```
Comment (cdietrich) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/preferences/BuilderConfigurationBlock.java]: OptionsConfigurationBlock is api so i wonder if we should make this method api too

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/preferences/OptionsConfigurationBlock.java [hunk 4]
```diff
@@ -581,6 +584,16 @@ private void logError(String text, IOException e) {
 		uiPlugin.getLog().log(status);
 	}
 
+	public Map<String, ValueDifference<String>> getBuildAffectingPreferenceChanges() {
+		Map<String, ValueDifference<String>> changes = getPreferenceChanges();
+		Map<String, ValueDifference<String>> buildAffectingChanges = Maps.filterEntries(changes, getAffectsBuilderPredicate());
+		return buildAffectingChanges;
+	}
+
+	protected Predicate<Entry<String, ValueDifference<String>>> getAffectsBuilderPredicate() {
```
Comment (cdietrich) [inline org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/preferences/OptionsConfigurationBlock.java]: OptionsConfigurationBlock is api so i wonder if we should make this method api too

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/preferences/BuilderConfigurationBlock.java (hunk 2 above)
Comment (szarnekow) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/preferences/BuilderConfigurationBlock.java]: I'd rather like to inject this one. Can we use a dedicated type here, though?

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/preferences/BuilderConfigurationBlock.java (hunk 3 above)
Comment (szarnekow) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/preferences/BuilderConfigurationBlock.java]: It should become API, yes. But I'd prefer a better type signature and easier means to customize it.

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/preferences/OptionsConfigurationBlock.java (hunk 4 above)
Comment (szarnekow) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/preferences/OptionsConfigurationBlock.java]: Should be make this a method `hasBuildAffectingPreferenceChanges()` that delegates to a utility?

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/preferences/BuilderConfigurationBlock.java (hunk 2 above)
Comment (szarnekow) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/preferences/BuilderConfigurationBlock.java]: Maybe a type with a method `isAffectingBuild(Map<String, ValueDiff<String>>)` would be easier to grasp and allow for easier customizing, too?

File: org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/XtextUiModule.java [hunk 5]
```diff
@@ -184,5 +186,8 @@ public void configureIResourceDescriptionsPersisted(com.google.inject.Binder bin
 	public void configureEditorScope(Binder binder) {
 		binder.bindConstant().annotatedWith(Names.named(XtextEditor.KEY_BINDING_SCOPE)).to("org.eclipse.xtext.xtext.ui.XtextEditorScope");
 	}
-	
+
```
Comment (cdietrich) [inline org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/XtextUiModule.java]: @tivervac why is this here and not in DefaultUiModule?

Comment (tivervac) [reply org.eclipse.xtext.xtext.ui/src/org/eclipse/xtext/xtext/ui/XtextUiModule.java]: Because I thought that was the place to put it, seems I was wrong ;)

**Comment by cdietrich** [comment]
@szarnekow any ideas how to resolve that without introducing a cycle?

**Comment by cdietrich** [comment]
or we use simply `@ImplementedBy`

**Comment by tivervac** [comment]
@cdietrich Using `@ImplementedBy` we still need to access to xtext.builder from xtext.ui, don't we? If so, that doesn't solve anything

**Comment by cdietrich** [comment]
dont get that: cant everything be moved to builder? what is needed in ui?

**Comment by cdietrich** [comment]
i am clueless. maybe the option and the builder page should be separated so that everything can be moved to builder

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/preferences/OptionsConfigurationBlock.java [hunk 6]
```diff
@@ -528,7 +531,7 @@ private void incrementRebuildCount() {
 	}
 
 	protected boolean processChanges(IWorkbenchPreferenceContainer container) {
-		boolean needsBuild = !getPreferenceChanges().isEmpty();
+		boolean needsBuild = buildPreferenceEvaluator.isAffectingBuild(getPreferenceChanges());
```
Comment (cdietrich) [inline org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/preferences/OptionsConfigurationBlock.java]: why is buider here in a generic class

Comment (cdietrich) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/preferences/OptionsConfigurationBlock.java]: maybe a simple method here and the delegate to the evaluator in subclass

Comment (tivervac) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/preferences/OptionsConfigurationBlock.java]: @szarnekow requested an injectable API for better customization

Comment (szarnekow) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/preferences/OptionsConfigurationBlock.java]: This is so much entangled, it's a shame, eg see this code below `container.registerUpdateJob(getBuildJob(getProject()));`

Bindings required for the builder are currently done by the Xtext generator, which I'd like to avoid.

It would be a safe bet if we moved the XtextBuildPreferenceEvaluator and defined the constant for the output config in ui and referenced it from `EclipseOutputConfigurationProvider.OUTPUT_PREFERENCE_TAG`

**Comment by tivervac** [comment]
@szarnekow Is this what you meant?

**Comment by szarnekow** [comment]
Yes exactly. What do you think? Is this a feasible approach?

**Comment by tivervac** [comment]
I think it's a bit nasty to declare those constants in UI instead of in the builder package but it's a lot better than generating it. Seems feasible to me.

**Comment by tivervac** [comment]
@szarnekow Is this ready to be merged now?

**Comment by cdietrich** [comment]
should be done after version bump

**Comment by szarnekow** [comment]
So the last remaining bit would be `@since 2.16`
